types/prismjs/index.d.ts

```diff
@@ -1,228 +1,435 @@
 }
 
-export type TokenNode = Token | string | Array<Token | string>;
+export type TokenNode = string | Token | Array<string | Token>;
 
 export class Token {
-    /**
-     * Given a token or string input, convert it to highlighted HTML
-     * @param type The token(s) to stringify
-     * @param language The language definition that understands the given token(s)
-     * @param parent The parent element element to pass into hook callbacks
-     * @returns The HTML that represents the given token(s)
-     */
-    static stringify(token: TokenNode, language: LanguageDefinition, parent: HTMLPreElement): string;
-
-    /**
-     * The constructor for the Prism Token class
-     * @param type The type of the token
-     * @param content The contents of the token
-     * @param alias Aliases for the current token (added as classes to this token's HTML)
-     * @param matchedStr The matched string that generated this token
-     * @param greedy Value that determines whether or not this token was generated using a greedy parsing algorithm
-     */
-    constructor(type: string, content: TokenNode, alias: any, matchedStr: string, greedy: boolean);
-
-    /** The type of the token */
-    type: string;
-
-    /** The content of the token */
-    content: TokenNode;
-
-    /** Other names for this token (added as classes to this token's HTML) */
-    alias: string;
+	/**
+	 * Creates a new token.
+	 *
+	 * @param type See {@link Prism.Token#type type}
+	 * @param content See {@link Prism.Token#content content}
+	 * @param [alias] The alias(es) of the token.
+	 * @param [matchedStr=""] A copy of the full string this token was created from.
+	 * @param [greedy=false] See {@link Prism.Token#greedy greedy}
+	 */
+	constructor(
+		type: string,
+		content: TokenNode,
+		alias?: string | string[],
+		matchedStr?: string,
+		greedy?: boolean
+	);
+
+	/**
+	 * The type of the token.
+	 *
+	 * This is usually the key of a pattern in a {@link Grammar}.
+	 */
+	type: string;
+
+	/**
+	 * The strings or tokens contained by this token.
+	 *
+	 * This will be a token stream if the pattern matched also defined an `inside` grammar.
+	 */
+	content: TokenNode;
+
+	/**
+	 * The alias(es) of the token.
+	 *
+	 * @see TokenObject
+	 */
+	alias: string | string[];
+
+	/**
+	 * The length of the matched string or 0.
+	 */
+	length: number;
+
+	/**
+	 * Whether the pattern that created this token is greedy or not.
+	 *
+	 * @see TokenObject
+	 */
+	greedy: boolean;
+
+	// tslint:disable:no-redundant-jsdoc-2
+	/**
+	 * Converts the given token or token stream to an HTML representation.
+	 *
+	 * The following hooks will be run:
+	 * 1. `wrap`: On each {@link Prism.Token}.
+	 *
+	 * @param token The token or token stream to be converted.
+	 * @param language The name of current language.
+	 * @param [parent] The parent token stream, if any.
+	 * @return The HTML representation of the token or token stream.
+	 * @private
+	 */
+	static stringify(
```

Comment:
The `Token.stringify(…)` method is private according to @RunDevelopment in https://github.com/PrismJS/prism/pull/1782#discussion_r261895197.

Reply:
I marked it as private because it can only work correctly if the tokens passed to it are already encoded via `Prism.util.encode`.
Put simply: It's a deadly trap from anyone trying to use it and it should only be used by the library anyway, so I went with private.

Reply:
In this case, maybe `Prism.util.encode` should return a different type which this function will accept instead of `TokenNode`?

Reply:
That wouldn’t work, since TypeScript uses structural equality when checking if two types match, and since the result of `Prism.util.encode("<x>")` is `"&lt;x&gt;"`, which is still `string`, it will match `TokenNode`.

Reply:
Something in the direction of `encode(tokens: TokenNode): EncodedTokenNode`, right?

Well, I don't know, if it's worth it as I plan to change this in the future anyway.

Reply:
There are certain ways to enforce nominal typing in TypeScript.

But even if you keep structural typing, having a different name for this type to differentiate between them would be a good idea IMHO.

Reply:
Except that type aliases are expanded into what they are aliases of in IDEs, so it wouldn’t really be clear from hovering over the function in **VS Code**.